flash loan function

dachen0 opened the pull request.

dachen0 committed:
being creating flash loan function

dachen0 committed:
add flash loan test

dachen0 committed:
add insufficient repay test case

dachen0 committed:
add flash loan events + remove some commented out stuff

dachen0 commented on contracts/suilend/sources/reserve.move:
```diff
@@ -953,6 +970,46 @@ module suilend::reserve {
+        // let borrow_amount_with_fees = amount + borrow_fee;
+
+        reserve.available_amount = reserve.available_amount - amount;
+        reserve.borrowed_amount = add(reserve.borrowed_amount, decimal::from(amount));
```
this is just amount borrowed but the repay func "repays" amount borrowed + fees. is this an issue? since it will always reduce reserve.borrowed_amount at the end of each flash loan cycle

0xLendlord replied:
I am thinking for safety we should consider the flash loan to Mutex lock the lending market. That way we prevent attack vectors that rely on sandwiching other lending market operations between the flash loan.

As for the the `reserve.borrowed_amount` it should remain the same after the flash repay, right

0xLendlord replied:
at borrow time:

reserve.available_amount -= amount_no_fees;
reserve.borrowed_amount += amount_with_fees;

at repay time:
reserve.available_amount += amount_with_fees
reserve.borrowed_amount -= amount_with_fees;

so borrowed_amount remains unchanged at the end and available_amount increases by the fee

dachen0 commented on contracts/suilend/sources/lending_market.move:
```diff
@@ -373,6 +392,51 @@ module suilend::lending_market {
+
+        let FlashLoan{ amount_to_repay } = flash_loan;
+        let repay_coins = coin::split(max_repay_coins, amount_to_repay, ctx);
+        reserve::repay_liquidity<P, T>(reserve, coin::into_balance(repay_coins), decimal::from(amount_to_repay));
```
see here that it repays the amount of repay coins which is amount borrowed + fees. I'm not sure what the correct way to handle fees should be.

0xLendlord commented on contracts/suilend/sources/lending_market.move:
```diff
@@ -64,6 +64,10 @@ module suilend::lending_market {
         obligation_id: ID,
     }
 
+    public struct FlashLoan {
```
Lets add coin type: `FlashLoan<T>` otherwise an attacker can spoof and repay the flash loan with a different coin

0xLendlord commented on contracts/suilend/sources/lending_market.move:
```diff
@@ -464,6 +562,27 @@ module suilend::lending_market {
+        assert!(reserve::coin_type(reserve) == type_name::get<T>(), EWrongType);
+
+        let (balance, amount_to_repay) = reserve::fulfill_flash_loan_liquidity_request(reserve, liquidity_request);
+        (
+            coin::from_balance(
+            balance,
+            ctx,
+            ),
+            FlashLoan {amount_to_repay}
+        )
```
```suggestion
        (
            coin::from_balance(
                balance,
                ctx,
            ),
            FlashLoan { amount_to_repay }
        )
```

0xLendlord commented on contracts/suilend/sources/lending_market.move:
```diff
@@ -373,6 +392,51 @@ module suilend::lending_market {
+
+        reserve::compound_interest(reserve, clock);
+
+        let FlashLoan{ amount_to_repay } = flash_loan;
```
```suggestion
        let FlashLoan { amount_to_repay } = flash_loan;
```

0xLendlord commented on contracts/suilend/tests/lending_market_tests.move:
```diff
@@ -651,6 +651,232 @@ module suilend::lending_market_tests {
+
+        // we can't actually test if the hot potato was dropped because
+        // the compiler won't let us use a "dropped" variable
+        // even though we know that if the tranasction failed *all* state
```
```suggestion
        // even though we know that if the transaction failed *all* state
```

0xLendlord committed:
flash loan refactor

0xLendlord commented on contracts/suilend/sources/lending_market.move:
```diff
@@ -447,6 +519,41 @@ module suilend::lending_market {
+            coin_type: type_name::get<T>(),
+            reserve_id: object::id_address(reserve),
+            liquidity_amount: liquidity_request.liquidity_request_amount(),
+            origination_fee_amount: liquidity_request.liquidity_request_fee(),
```
rename `flash_loan_fee_amount`

0xLendlord replied:
renamed to `fee_amount` as the flash loan part is self evident

0xLendlord commented on contracts/suilend/sources/lending_market.move:
```diff
@@ -447,6 +519,41 @@ module suilend::lending_market {
+            lending_market_id,
+            coin_type: type_name::get<T>(),
+            reserve_id: object::id_address(reserve),
+            liquidity_amount: liquidity_request.liquidity_request_amount(),
```
maybe amount without fee?

0xLendlord replied:
renamed to `borrow_amount` as is now net of fees

0xLendlord commented on contracts/suilend/sources/lending_market.move:
```diff
@@ -447,6 +519,41 @@ module suilend::lending_market {
+            assert!(amount > 0, ETooSmall);
+        };
+
+        let liquidity_request = reserve::borrow_liquidity<P, T>(reserve, amount);
```
flash loan fee should be a separate fee.

0xLendlord replied:
done

0xLendlord committed:
- added fee to flash borrow events
- borrow amount field in events is net of fees
- added fee field to flash loan hot potato
- flash loan specific fee
- tests

0xLendlord committed:
rename flash loan events borrow amount field to liquidity amount

0xLendlord committed:
Merge remote-tracking branch 'origin/devel' into flash-loan

0xLendlord committed:
liquidity_amount in event is gross of fees
